lib/Sema/CSGen.cpp

```diff
@@ -3101,6 +3122,18 @@ namespace {
     Type visitKeyPathDotExpr(KeyPathDotExpr *E) {
       llvm_unreachable("found KeyPathDotExpr in CSGen");
     }
+    
+    Type visitTapExpr(TapExpr *expr) {
+      auto locator = CS.getConstraintLocator(expr);
+      auto tv = CS.createTypeVariable(locator);
+
+      if (auto subExpr = expr->getSubExpr()) {
+        auto subExprType = CS.getType(subExpr);
+        CS.addConstraint(ConstraintKind::Bind, subExprType, tv, locator);
```

Comment:
One small change which I think might work here - instead of biding sub-expr type to type variable, you can return `subExprType` directly and allocate type variable only if there was no sub-expression...

Reply:
Lower priority than the runtime performance stuff, but I'll look at it. Thanks!